File: Lab3_19/2trap.cpp
// Найти численно значение неопределенного интеграла exp(-2*x) * sin(19*x) с точностью до 10^-6. Метод трапеций.
#include <iostream>
#include <cmath>
#include <iomanip>

using namespace std;

double f(double x) {
    return exp(-2*x) * sin(19*x); // Функция для интегрирования
}

double integrate(double a, double b, double tolerance) {
    int n = 1;
    double h = (b - a) / n;
    double integral = 0.5 * (f(a) + f(b)); // Первое приближенное значение интеграла
    double integral_prev = 0;
    
    while (abs(integral - integral_prev) >= tolerance) {
        integral_prev = integral;
        double sum = 0;
        
        for (int i = 1; i <= n - 1; ++i) {
            sum += f(a + i * h);
        }
        
        integral = 0.5 * integral + h * sum;
        n *= 2;
        h = (b - a) / n;
    }
    
    return integral / 2;
}

int main() {
    double a = 0.0; // Пределы интегрирования 
    double b = 1.0;
    double tolerance = 1e-6; // Задаем точность

    double result = integrate(a, b, tolerance);
    
    cout << "Метод трапеций"<< endl;
    cout << "Интегрируемая функция:(1,2) f(x) = exp(-2*x) * sin(19*x)"<<endl;
    cout << "Значение определенного интеграла с точностью 10^-6: " << endl;
    cout << "I = " << fixed << setprecision(12) << result << endl;

    return 0;
}

File: Lab3_19/2trap.py
#Найти численно значение неопределенного интеграла exp(-2*x) * sin(19*x) c точностью 10-6с. Метод трапеций с заданием точности
def trapezoid_rule(func, a, b, nseg):
    """Правило трапеций
       nseg - число отрезков, на которые разбивается [a;b]"""
    dx = 1.0 * (b - a) / nseg
    sum = 0.5 * (func(a) + func(b))
    for i in range(1, nseg):
        sum += func(a + i * dx)

    return sum * dx

def func(x):
    return exp(-2*x) * sin(19*x) 

from math import *

print("Используем метод трапеций")
print("Интегрируемая функция:(x) = e^(-2x) * sin(19x)")


a=0 # левый край отрезка интегрирования
b=1 # правый край отрезка интегрирования
nseg = 2 # начальное количество сегментов
eps=1e-6 # точность


print('a = ', a ,', b = ', b ,', eps = ', eps, sep='')

int_1 = trapezoid_rule(func, a, b, nseg)
nseg *= 2
int_2 = trapezoid_rule(func, a, b, nseg)

while abs(int_1 - int_2) > eps:
    nseg *= 2
    int_1 = trapezoid_rule(func, a, b, nseg)
    nseg *= 2
    int_2 = trapezoid_rule(func, a, b, nseg)

print("\nОтвет: I =", int_2, "\nКоличество разбиений:", nseg)